Consistent configuration properties

potiuk opened the pull request:
Make sure you have checked _all_ steps below and that above description is correct.

### JIRA
- [x] My PR addresses the following [Issue](https://github.com/GoogleCloudPlatform/oozie-to-airflow/issues/178)
issue and references it in commit message.

### Tests
- [x] My PR has unit tests testing the functionality (or I explained why it needs no tests)

### Commits
- [x] My commits reference the issue in the message using: Fixes #XXX /Closes #XXX or similar.

### Documentation
- [x] My PR adds documentation that describes how to use it (or I explained why it needs no docs)

mik-laj commented on o2a/o2a_libs/property_utils.py:
```diff
@@ -0,0 +1,69 @@
+    ):
+        self.configuration_properties: Dict[str, str] = configuration_properties
+        self.job_properties: Dict[str, str] = job_properties
+        self.action_node_properties: Dict[str, str] = action_node_properties if action_node_properties else {}
```
```suggestion
        self.action_node_properties: Dict[str, str] = action_node_properties or {}
```

potiuk replied:
👍

mik-laj commented on o2a/mappers/action_mapper.py:
```diff
@@ -24,22 +25,45 @@
+            oozie_node, name, dag_name, job_properties, configuration_properties, trigger_rule, **kwargs
+        )
+        self.action_node_properties: Dict[str, str] = {}
+        self._parse_config()
```
I think that this should be done while parsing the node. What do you think about moving this call to `on_parse_node`?

potiuk replied:
Yes. but only if it is done in the class and _on_parse_node uses super(). I think all logic related to config parsing should be done in the ActionMapper not in the deriving classes. I will propose a change for that.

mik-laj replied:
Please look at: https://github.com/GoogleCloudPlatform/oozie-to-airflow/pull/166/files

potiuk replied:
Proposal here: 7e489b44fd6d74850ff3768bb5acc0b6c950b46b

potiuk replied:
Please mark it as resolved @mik-laj  if you think my proposal is OK.

mik-laj commented on o2a/mappers/action_mapper.py:
```diff
@@ -24,22 +25,45 @@
         config = self.oozie_node.find("configuration")
         if config:
+            property_set = self.property_set
             property_nodes = xml_utils.find_nodes_by_tag(config, "property")
```
This logic is repeated in several places.I think we should create an helper method - extract_properties_from_configuration. We can do it in seperate PR.

potiuk replied:
No need for that. In fact it was exactly repeated in one place only - in subworkflow_mapper which was completely not needed as it already derives from action mapper.

I removed it in 554a0d9bf56df6123891998b81df35b7684b106f

mik-laj commented on o2a/mappers/dummy_mapper.py:
```diff
@@ -14,16 +14,37 @@
+    """Dummy mapper used in place of not-yet-implemented mappers """
+
+    def __init__(self, oozie_node: Element, name: str, dag_name: str, **kwargs):
+        kwargs, job_properties = remove_key_from_dictionary_copy(kwargs, "job_properties", {})
```
I do not understand the purpose of this code

potiuk replied:
The problem was that I wanted to keep configuration/job properties obligatory for all mappers - except kill and dummy mappers. Since we should not use = {} in parameter default values (mutable dict) I tried to use = None before and override it with {} inside but mypy complained about the Optional[Dict]. However I found a simpler way of doing it - see e0ff6f1bf016cd3a42d7e3f57dedd40aade8f748

potiuk replied:
It actually fixed even better in e4afbfddee0b8e9bdea5ff11f797a06b523d39a8

mik-laj commented on o2a/mappers/fs_mapper.py:
```diff
@@ -112,47 +116,85 @@ class FsMapper(ActionMapper):
-    def parse_fs_action(self, index: int, node: Element):
+    def parse_fs_action(self, index: int, node: Element) -> Optional[Task]:
         tag_name = node.tag
+        if tag_name == "configuration":
```
I have a different idea. I think we should omit the nodes here, which are not supported. ie. parse only supset of nodes
```
            <xs:choice minOccurs="0" maxOccurs="unbounded">
                <xs:element name="delete" type="workflow:DELETE"/>
                <xs:element name="mkdir" type="workflow:MKDIR"/>
                <xs:element name="move" type="workflow:MOVE"/>
                <xs:element name="chmod" type="workflow:CHMOD"/>
                <xs:element name="touchz" type="workflow:TOUCHZ"/>
                <xs:element name="chgrp" type="workflow:CHGRP"/>
            </xs:choice>
```
Therefore, the configuration will normally be supported by ActionMapper. 

ActionMapper should have a code similar to this:
```
class ActionMapper(BaseMapper):
    """Base class for all action mappers"""

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.properties: Dict[str, str] = {}

    def on_parse_node(self, workflow):
        super().on_parse_node(workflow)
        self._parse_configuration(workflow)
```

potiuk replied:
👍 Fixed in ccc01c1853457a0bec5ab0cef8e088b5b2d4ae54

mik-laj commented on o2a/mappers/spark_mapper.py:
```diff
@@ -47,60 +47,63 @@ def __init__(
 
-    @staticmethod
-    def _get_or_default(root: ET.Element, tag: str, default: str = None, params: Dict[str, str] = None):
+    def _get_or_default(self, root: ET.Element, tag: str, default: str = None):
```
This method looks very similar to `get_tag_el_text`?

potiuk replied:
True! Fixed in e8cd0f510d38afcd96da054d70f4fd61037f20d0

mik-laj commented on o2a/mappers/ssh_mapper.py:
```diff
@@ -37,48 +38,68 @@ def __init__(
+        return cmd
 
-        self.command = el_utils.convert_el_to_jinja(cmd, quote=True)
+    def get_host_key(self, property_set: PropertySet) -> str:
```
Is this argument necessary? I think this method can refer to this object directly.

potiuk replied:
Indeed. Fixed in 46e3cd5b8e2835bc191622530affa0f5a5e54a0b

mik-laj commented on o2a/mappers/subworkflow_mapper.py:
```diff
@@ -99,26 +105,26 @@ def _parse_config(self):
 
-    def to_tasks_and_relations(self):
-        tasks = [
+    def to_tasks_and_relations(self) -> Tuple[List[Task], List[Relation]]:
```
The parent class has type definitions. Is this change necessary?

potiuk replied:
I think it's much nicer to add it here for potential reader in the future. This way you do not have to look at the base class to see what types should be returned. And mypy will complain if those will differ. So I'd rather leave it here.

Please resolve it @mik-laj  if you are ok with it.

mik-laj commented on o2a/templates/workflow.tpl:
```diff
@@ -12,16 +12,25 @@
+}
+CONFIGURATION_PROPERTIES={
+   {% for key, value in configuration_properties.items() %}
+       '{{key | python_escape}}': {% if value is none %}None{% else %}'{{ value | python_escape }}'{% endif %},
```
What do you think about adding None to the filter?

mik-laj replied:
List is also be usefull.

potiuk replied:
I'd rather do it in the future but I agree it's a very good idea to simplify the templates a lot!

potiuk replied:
As discussed - we will do it in the future

mik-laj commented on o2a/converter/oozie_converter.py:
```diff
@@ -94,14 +95,20 @@ def __init__(
             input_directory_path=input_directory_path,
             output_directory_path=output_directory_path,
-            params=params,
+            job_properties=self.job_properties,
```
Why did not you pass the PropertySet object here?

potiuk replied:
Good point. Will do.

potiuk replied:
Also I will do it in ActionMapper and BaseMapper!

potiuk replied:
It will simplify the code quite a lot if I do it this way it seems.

potiuk replied:
Fixed in e0ff6f1bf016cd3a42d7e3f57dedd40aade8f748

mik-laj commented on o2a/mappers/dummy_mapper.py:
```diff
@@ -14,16 +14,37 @@
+            **kwargs,
+        )
+
+    def to_tasks_and_relations(self) -> Tuple[List[Task], List[Relation]]:
```
The parent class contains type definitions. Is it necessary?

potiuk replied:
Again - I prefer to repeat it for readability

mik-laj commented on o2a/mappers/null_mapper.py:
```diff
@@ -13,19 +13,17 @@
 
     # pylint: disable=no-self-use
-    def to_tasks_and_relations(self):
+    def to_tasks_and_relations(self) -> Tuple[List[Task], List[Relation]]:
```
I think that we should delete this mapper. It is useless. He is useless. When it is used it will always generate an incorrect DAG, because the relationship is not created correctly.

potiuk replied:
Removed in 6cedebe6a5b0c363fcae1bb4af54edf2309d8d7f

mik-laj commented on o2a/mappers/prepare_mixin.py:
```diff
@@ -27,24 +29,25 @@ class PrepareMixin:
+        delete = None
+        mkdir = None
+        if not delete_paths and not mkdir_paths:
+            raise Exception(f"There is neither delete nor mkdir in the {oozie_node}'s prepare.")
```
In XSD schema, I do not see such a limitation. In automatically generated XML files, this element can therefore appear. I think this check should appear in the has_prepare method. What do you think?

potiuk replied:
I actually simplified it even further. I reworked prepare_mixin to take oozie_node as __init__ parameter and the whole prepare functionality now is pretty much embedded in Prepare mixin.  See 40e3e4d4ec448c28f2ff4c60e5384ca20083ed23

potiuk replied:
please resolve it @mik-laj if you think the discussion is resolved.

mik-laj commented on o2a/mappers/shell_mapper.py:
```diff
@@ -35,48 +35,61 @@ def __init__(
-                task_id=self.name + "_prepare",
-                template_name="prepare.tpl",
-                template_params=dict(prepare_command=prepare_command),
+    def to_tasks_and_relations(self) -> Tuple[List[Task], List[Relation]]:
```
The parent class have a type defintion. It is neccesary?

potiuk replied:
Same as above. I prefer to leave it for reader's clarity..

potiuk replied:
Please @mik-laj resolve this if you think the discussion is resolved.

mik-laj commented on o2a/mappers/start_mapper.py:
```diff
@@ -13,16 +13,33 @@
 class StartMapper(BaseMapper):
+    """Maps start node"""
+
+    def __init__(self, oozie_node: Element, name: str, dag_name: str):
```
What is purpose of this code? I think, It useless, because BaseMapper provide a good constructor.

potiuk replied:
It was to have dummy trigger_rule. But after your comment I found that it will be even easier to simplify the whole mapper creation instead and hard-code the rule when the Start Mapper is created int that one place it was used. On the way I also removed CONTROL_MAP map for the control mappers as it turned out to be completely not needed with this change. See abb241def467a5cbef041ebe6c78d33a7c97541e

potiuk replied:
Please @mik-laj resolve this is if you think the discussion is resolved.

mik-laj commented on o2a/utils/template_utils.py:
```diff
@@ -18,13 +18,17 @@
 TEMPLATE_CACHES: Dict[str, Any] = {}
 
 TEMPLATE_ENV.filters["to_var"] = convert_to_python_variable
+TEMPLATE_ENV.filters["python_escape"] = escape_string_with_python_escapes
```
I think, this filter should return a text with quote. This will make its use much easier. What do you think about taking this change into a separate PR?

potiuk replied:
Yep. We could probably implement it together with implementing the generic filter for python_escape.

mik-laj replied:
I started thinking about it. I think that we should not completely implement function change at all. It seems to me that it is not necessary to introduce a different way of formatting strings in this PR. At the same time losing a lot of functionality that filter to_json provides. to_json is good enough in this case.  If we introduce the current version of this filter, it will be more difficult to restore the old functionality.  I think it is worth doing it once and well.

Let's look at how it will look. We assume that the variable ARG_TEXT contains string and the variable ARG_LIST contains list of string

1. 
```python
AwesomeStuff(
   arg_text={{ arg_text | to_json }},
   {% if arg_list %}
   arg_list={{ arg_list | tojson }},
   {% endif %}
)
```
2. 
```python
AwesomeStuff(
    arg_text={{ arg_text | to_json }},
     {% if arg_list %}arg_list=[
          {% for arg in arg_list %}
            '{{ arg | python_escape }}',
        {% endfor %}],
    {% endif %}
)
```
3. 
```python
AwesomeStuff(
   arg_text={{ arg_text | python_escape }},
   {% if arg_list %}
   arg_list={{ arg_list | python_escape }},
   {% endif %}
)
```
I believe that the introduction of new functionality should be at least equivalent to the current solution. I believe that the introduction of new functionality should be as valuable as the current solution, or more. Introducing this filter in this version in this change will bring us a lot more work in the future, which will be easier to do if this change is not included.

potiuk replied:
I think I have found a really good solution to that.

I agree that moving '' into the escape function is a good idea. But I think using a generic filter (like tojson) is very misleading because the reader does not know the intention. I reviewed all the usages of the filter so far and it seems that we only have three cases: string, list of strings and dict of string, string. In all cases the strings are optional and can be turned to None. So far we have no bool, int nor others and we always know what type we are going to escape. Therefore I think much simpler and explicit solution is to implement three filters - python_escape_string, python_escape_list, python_escape_dictionary and use those filters as appropriate. 

I implemented it in this fixup: 4fd30e46bfda45bd80b44a75f36279caee03008c 

Let me know if @mik-laj if that resolves the discussion.

potiuk replied:
I am also open to discuss adding a wrapper "filter" to detect the type and choose one of those three (or more in the future) but I would like to discuss it in the team as I think separate filters express the intention much better.

mik-laj replied:
I prepared my proposition at https://github.com/GoogleCloudPlatform/oozie-to-airflow/pull/224

mik-laj commented on o2a/utils/template_utils.py:
```diff
@@ -18,13 +18,17 @@
 TEMPLATE_LOADER = jinja2.FileSystemLoader(searchpath=TPL_PATH)
-TEMPLATE_ENV = jinja2.Environment(loader=TEMPLATE_LOADER, undefined=jinja2.StrictUndefined)
+TEMPLATE_ENV = jinja2.Environment(
+    loader=TEMPLATE_LOADER, undefined=jinja2.StrictUndefined, trim_blocks=True, lstrip_blocks=True
```
What do you think about taking this change into a separate PR?

potiuk replied:
I needed that in order to make sure the jinja templates with proper configuration values is rendered as needed. I think it will be very difficult to split it off. 

If you know how to do it and can help with that I will be greatful.

potiuk replied:
Please resolve this if you think it resolves the discussion @mik-laj

mik-laj commented on tests/mappers/test_archive_extractor.py:
```diff
@@ -18,38 +18,46 @@
         archive_extractor.add_archive("test_archive.zip")
         # Then
-        self.assertEqual(archive_extractor.archives, ["test_archive.zip"])
+        self.assertEqual(["test_archive.zip"], archive_extractor.archives)
```
What do you think about taking this change into a separate PR/commit?  The current commit is very large.

potiuk replied:
It will not change the size of the commit heavily. What exactly do you propose to split :)?

mik-laj replied:
Make it easy to make reviews, but it's not required.

mik-laj commented on tests/mappers/test_fs_mapper.py:
```diff
@@ -24,12 +25,13 @@
 
-# pylint: disable=invalid-name
-class prepare_mkdir_commandTest(unittest.TestCase):
+class PrepareMkdirCommandTest(unittest.TestCase):
```
This makes it difficult to find the test code using the search.

potiuk replied:
The naming was wrong. What you should do instead is to name the test methods appropriately rather than classes. I renamed the methods properly now and they are easily searchable.
See c8bdaebb1674b8e9f2d07437286668e8d436ef9b

potiuk replied:
I improved it even further. I moved all the prepare_<>_command tests in single class and now it is much better conforming to Python standards (and we have no unnecessary classes created).
Please resolve this @mik-laj if you think it resolves the discussion.

sprzedwojski commented on o2a/converter/oozie_converter.py:
```diff
@@ -136,11 +143,24 @@ def convert_nodes(nodes: Dict[str, ParsedActionNode]):
-    def add_properties_to_params(self, params: Dict[str, str]):
+    def read_configuration_properties(self):
+        """
+        Reads configuration job_properties to configuration_properties dictionary.
```
```suggestion
        Reads configuration configuration_properties to configuration_properties dictionary.
```

sprzedwojski commented on o2a/converter/oozie_converter.py:
```diff
@@ -136,11 +143,24 @@ def convert_nodes(nodes: Dict[str, ParsedActionNode]):
+    def read_configuration_properties(self):
+        """
+        Reads configuration job_properties to configuration_properties dictionary.
+        Replaces EL job_properties within.
```
```suggestion
        Replaces EL configuration_properties within.
```

potiuk replied:
That was a bad rename properties -> job_properties at some point of time. Resolved in 3d1455955ce0a10df49cf13602d02696f6c89318

mik-laj commented on o2a/mappers/action_mapper.py:
```diff
@@ -24,22 +25,45 @@
 # pylint: disable=abstract-method
 # noinspection PyAbstractClass
-class ActionMapper(BaseMapper):
+from o2a.o2a_libs.property_utils import PropertySet
```
@potiuk Can you look at this?

potiuk replied:
Fixed in 2bc950a43ecb5faba2f16ccfd4c50372ccd7db61

mik-laj commented on o2a/mappers/prepare_mixin.py:
```diff
@@ -13,38 +13,48 @@
-        delete_paths, mkdir_paths = self.parse_prepare_node(oozie_node, params)
-        if delete_paths or mkdir_paths:
+    def has_prepare(self):
+        prepare_nodes = xml_utils.find_nodes_by_tag(self.oozie_node_for_prepare, "prepare")
```
````
<xs:element name="prepare" type="workflow:PREPARE" minOccurs="0" maxOccurs="1"/>
````
I think, you should use `find_node_by_tag`.

potiuk replied:
Indeed. a bit simpler. Fixed in 635479a79f2413a4a0298e4878d062c9b4d72ccf

mik-laj commented on o2a/mappers/prepare_mixin.py:
```diff
@@ -13,38 +13,48 @@
+            delete_nodes = xml_utils.find_nodes_by_tag(node, "delete")
+            mkdir_nodes = xml_utils.find_node_by_tag(node, "mkdir")
+            if delete_nodes or mkdir_nodes:
+                _has_prepare = True
```
```suggestion
               return True
```

potiuk replied:
Fixed in f80387c

mik-laj commented on o2a/mappers/prepare_mixin.py:
```diff
@@ -13,38 +13,48 @@
+            mkdir_nodes = xml_utils.find_node_by_tag(node, "mkdir")
+            if delete_nodes or mkdir_nodes:
+                _has_prepare = True
+        return _has_prepare
```
return False

potiuk replied:
Indeed. Fixed in f80387cfcf7b7b60919ac025449d3640bc19c218

mik-laj commented on o2a/mappers/prepare_mixin.py:
```diff
@@ -13,38 +13,48 @@
+        delete_paths, mkdir_paths = self.parse_prepare_node(property_set=property_set)
+        if not delete_paths and not mkdir_paths:
+            return None
+        delete = None
```
```suggestion
        delete = " ".join(delete_paths) if delete_paths else None 
```
This makes it easier to read the code.

potiuk replied:
True. Fixed in 889d35e7beccee5306f99cb079b2c860540fa9c0

mik-laj commented on tests/mappers/test_subworkflow_mapper.py:
```diff
@@ -71,23 +70,30 @@ def tearDown(self) -> None:
+                "nameNode": "hdfs://",
+                "oozie.wf.application.path": "hdfs:///user/pig/examples/pig",
+                "resourceManager": "localhost:8032",
+                "user.name": "potiuk",
```
I think, we should avoid store similar information in tests. This can be confused with real data.

potiuk replied:
True. Fixed in f475ebb0a04878d2efa25d171d6a0b0e569af5ff

potiuk commented:
@mik-laj -> please review all the unresolved ones - we will discuss it after the standup today and hopefully we can reach some conclusions.

mik-laj commented on tests/mappers/test_subworkflow_mapper.py:
```diff
@@ -78,16 +78,11 @@ def test_create_mapper_jinja(self, parse_els):
+        self.assertEqual("hdfs://", mapper.property_set["nameNode"])
+        self.assertEqual("hdfs:///user/pig/examples/pig", mapper.property_set["oozie.wf.application.path"])
+        self.assertEqual("localhost:8032", mapper.property_set["resourceManager"])
+        self.assertIsNotNone(mapper.property_set["user.name"])
```
Do you think it's worth adding a test to check the keys in the dictionary?  Now a new key may appear, which will not be checked in any test.

potiuk replied:
I think so. We are checking specific mapper as an example, so we can expect that those properties are in the property_set.

mik-laj commented on o2a/mappers/fs_mapper.py:
```diff
@@ -112,47 +116,79 @@ class FsMapper(ActionMapper):
+        tasks: List[Task] = []
+        index = 0
+        for node in self.oozie_node:
+            task = self.parse_fs_action(index, node)
```
Maybe it is better to prevent such a case by moving the condition check earlier?

mik-laj replied:
I have a ready patch that improves it.

mik-laj replied:
My patch is available: 
https://github.com/GoogleCloudPlatform/oozie-to-airflow/commit/d2691556b3c8b5840b626c813b6c474277c17ba1

sprzedwojski replied:
Maybe instead of such patches we should add a commit to the PR itself? I think it would be clearer and easier to review.

mik-laj replied:
I prefer not to change this PR without the approval of Jarek.

potiuk replied:
Let's discuss one-by-one and reach consensus what we do. I plan to squash and possibly re-split the change to multiple commits to separate out some changes so we can do whatever we decide as appropriate.

mik-laj commented on o2a/mappers/fs_mapper.py:
```diff
@@ -112,47 +116,79 @@ class FsMapper(ActionMapper):
+        if tag_name not in FS_OPERATION_MAPPERS.keys():
+            # Skip nodes that are not FS mapper action nodes
+            return None
         tasks_count = len(self.oozie_node)
```
This assumption is now incorrect. Nodes can be more than operations.

mik-laj replied:
My patch is available: 
https://github.com/GoogleCloudPlatform/oozie-to-airflow/commit/d2691556b3c8b5840b626c813b6c474277c17ba1

potiuk replied:
I left the discussion above open.

mik-laj commented on o2a/mappers/spark_mapper.py:
```diff
@@ -170,39 +141,21 @@ def _get_tasks(self):
                 arguments=self.application_args,
-                archives=self.hdfs_archives,
-                files=self.hdfs_files,
+                hdfs_archives=self.hdfs_archives,
```
I think you should change the name of this variable to match its purpose in the template. 
```bash
sed -i "/hdfs_archive/archive/" o2a/mappers/spark_mapper.py
```

potiuk commented:
This is a WIP yet - partially done splitting into separate commits

potiuk commented:
@sprzedwojski @mik-laj - > looking forward to your review :)

potiuk commented on tests/converter/test_oozie_parser.py:
```diff
@@ -38,7 +38,7 @@ def setUp(self):
             params=params,
             action_mapper=ACTION_MAP,
-            control_mapper=CONTROL_MAP,
+            dag_name="BBB",
```
I was following "AAA" dag name which was already in the code. I change both now :)

potiuk replied:
Fixed in c331a75265e5d20b3648e062f007714cf0f1573b

potiuk commented on o2a/o2a_libs/property_utils.py:
```diff
@@ -0,0 +1,69 @@
+    """
+
+    def __iter__(self) -> Iterator[str]:
+        return self.job_properties_merged.__iter__()
```
I did not like it either. I reworked it a bit now and possibly made more straightforward. no [] implementation for PropertySet - we are using now explicit params.props.merged[] to access the merged properties. Explicit is better than implicit. I also shortened the names to make the generated code less verbose. I think it should be much clearer now.

potiuk replied:
@sprzedwojski  -> please take a look at b652df97d111074a868cdcb05b3711a61d61cbf8 and see if it looks right.  Maybe we can come up with better name than "merged" ?

Using properties in templates:
```
arguments=[
      "{{ params.props.merged['mapreduce.input.fileinputformat.inputdir'] }}",
      "{{ params.props.merged['mapreduce.output.fileoutputformat.outputdir'] }}",
],
```

Definition of properties at the top of DAG:
```
CONFIG = {
    "dataproc_cluster": "oozie-51",
    "gcp_conn_id": "google_cloud_default",
    "gcp_region": "europe-west3",
    "gcp_uri_prefix": "gs:///dags",
}

JOB_PROPS = {
    "user.name": "potiuk",
    "nameNode": "hdfs://",
    "resourceManager": "localhost:8032",
    "queueName": "default",
    "examplesRoot": "examples",
    "oozie.use.system.libpath": "true",
    "userName": "potiuk",
    "oozie.wf.application.path": "/user/potiuk/examples/apps/demo",
}

Using config properties:
...
    pig_node_prepare = bash_operator.BashOperator(
        task_id="pig-node_prepare",
        trigger_rule="all_success",
        bash_command="$DAGS_FOLDER/../data/prepare.sh "
        "-c {{ params.config['dataproc_cluster'] }}"
        "-r {{ params.config['gcp_region'] }}"
        "-d %s"
        "-m %s"
        % (
            shlex.quote("/user/potiuk/examples/apps/demo/pig/output-data"),
            shlex.quote("/user/potiuk/examples/apps/demo/pig/created-folder"),
        ),
        params=PropertySet(config=CONFIG, job_properties=JOB_PROPS),
    )
```

passing Dictionary of merged job/action properties directly to the operator:

```
        dataproc_pig_properties=PropertySet(config=CONFIG, job_properties=JOB_PROPS).merged,

```

overriding merged properties with action-node-specific properties (it's not used here but could be via params.merged['mapred.job.queue.name']
```
    shell_node = bash_operator.BashOperator(
        task_id="shell-node",
        trigger_rule="dummy",
        bash_command="gcloud dataproc jobs submit pig "
        "--cluster={{params.config['dataproc_cluster']}} "
        "--region={{params.config['gcp_region']}} "
        "--execute %s" % shlex.quote("sh 'java -version'"),
        params=PropertySet(
            config=CONFIG,
            job_properties=JOB_PROPS,
            action_node_properties={"mapred.job.queue.name": "default"},
        ),
    )
```

potiuk replied:
@sprzedwojski -> please take a look at b652df9 and see if it looks right

potiuk commented on o2a/converter/oozie_converter.py:
```diff
@@ -133,11 +142,24 @@ def convert_nodes(nodes: Dict[str, ParsedActionNode]):
+            properties_file=self.configuration_properties_file, property_set=self.property_set
+        )
+
+    def update_job_properties(self):
```
I think read_and_update_job_properties_replace_el might be even better. the essence of this method is that it reads the properties from the file and updates the existing properties with those read from the file. The fact that we are replacing EL properties as well is really the "read" part (read property file but replace

potiuk replied:
I think this should be better: 70d2f965146795f65c603085246fc7286ea899e5 . Please Resolve @sprzedwojski if you think it's ok :)

potiuk replied:
Please double check if 70d2f96 is solving the problem better @sprzedwojski

potiuk commented on o2a/o2a_libs/property_utils.py:
```diff
@@ -0,0 +1,69 @@
+# pylint: disable=too-few-public-methods
+class PropertySet(Mapping):
+    """Holds all the different types of properties (job/action node for now - job.xml and workflow.xml in
+       the future) and implements [] operator to return property value according to the oozie algorithm
```
Fixed in b652df97d111074a868cdcb05b3711a61d61cbf8

potiuk commented:
@sprzedwojski  -> correction after review applied. 

Also @sprzedwojski + @mik-laj I added one more commit (it's sort of experimental but I thought I give it a try). We discussed it with Kamil that PrepareMixin was not really good "mixin" case. It was also leaking abstraction because it was necessary to use BaseMapper fields if you wanted to extract all prepare logic to the mixin - otherwise the mappers had to implement the same prepare-related logic (inserting prepare task, first_task_id) on their own.

I changed it in  5d4466d36f337e7d8f002ef01945b9285176e7da to use composition rather than mixin-inheritance and I feel much better about it. I named it PrepareMapperExtension, it has all the prepare logic (no logic in the mappers!). It takes BaseMapper as explicit init parameter (so coupling is now explicit rather than implicit as it was before). 

We might or might not want to use it now and merge together (I can separate it out completely). But maybe it's worth adding it since It makes the code quite a bit cleaner. To be discussed at planning/standup.

mik-laj commented on o2a/mappers/git_mapper.py:
```diff
@@ -63,52 +65,53 @@ def __init__(
+    def to_tasks_and_relations(self) -> Tuple[List[Task], List[Relation]]:
+        action_task = Task(
+            task_id=self.name,
+            template_name="git.tpl",
```
It looks like there are no tests for this template. I tried to find the change in the tests corresponding to the change and there is no test class

potiuk replied:
Good catch! I will add it as separate case

potiuk replied:
You already added it. Merged

mik-laj commented on o2a/templates/decision.tpl:
```diff
@@ -12,27 +12,26 @@
+{% for key, val in case_dict.items() %}
+{% if loop.first %}
     if {{ key }}:
         return "{{ val }}"
```
```suggestion
        return {{ val | to_python }}
```

potiuk replied:
You've added it  apparently merged.

mik-laj commented on o2a/templates/decision.tpl:
```diff
@@ -12,27 +12,26 @@
-{%- if not loop.first and not loop.last %}
+{% if not loop.first and not loop.last %}
     elif {{ key }}:
         return "{{ val }}"
```
```suggestion
        return {{ val | to_python }}
```

mik-laj commented on o2a/templates/decision.tpl:
```diff
@@ -12,27 +12,26 @@
-{%- if loop.last %}
+{% if loop.last %}
     else:
         return "{{ val }}"
```
```suggestion
        return {{ val | to_python }}
```

potiuk replied:
You added it. merged.

mik-laj commented on o2a/utils/el_utils.py:
```diff
@@ -182,8 +189,9 @@ def comma_separated_string_to_list(line: str) -> Union[List[str], str]:
-    url_with_var = replace_el_with_var(url, params=params, quote=False)
+def normalize_path(url, props: PropertySet, allow_no_schema=False):
+    url_with_var = replace_el_with_var(url, props=props, quote=False)
+    url_with_var = replace_el_with_var(url_with_var, props=props, quote=False)
```
I do not understand the purpose of this change.

potiuk replied:
That was bad merge : fixed in 4a956b1daef8318aea49d8561a77cc04a22e4c4b

mik-laj commented on o2a/templates/prepare.tpl:
```diff
@@ -12,10 +12,24 @@
+ However we can read from ~/data -> /home/airflow/gcs/data.
+ The easiest way to access it is using the $DAGS_FOLDER env variable.
+#}
+    bash_command="$DAGS_FOLDER/../data/prepare.sh "
```
I think this should be in the `prepare_command.tpl file to keep one style.

potiuk replied:
Good idea. Split in 81758e1c3b33e048bc1b4b471ca1e52818b4d451

potiuk commented on o2a/mappers/extensions/prepare_mapper_extension.py:
```diff
@@ -0,0 +1,93 @@
+                    raise Exception(f"Unknown XML node in prepare: {node.tag}")
+        return delete_paths, mkdir_paths
+
+    def insert_prepare_if_needed(
```
I thought about it a bit I think you are quite right. I rewrote it slightly - I split this method into two. Indeed prepare_mapper_extension only returns prepare task and I added prepend_task to ActionMapper so that it can be used in the mappers. Now there is very little code duplication between mappers and we have single responsibility of the mapper extension. See 5290f95481ad4726cd3f97329f396541977f85de

potiuk replied:
Please let me know @sprzedwojski if that's good :)

sprzedwojski replied:
I really like it now :+1:

potiuk commented on o2a/mappers/fs_mapper.py:
```diff
@@ -112,47 +116,75 @@ class FsMapper(ActionMapper):
+        tasks: List[Task] = []
+        valid_operation_tags = FS_OPERATION_MAPPERS.keys()
+        operation_nodes = [node for node in self.oozie_node if node.tag in valid_operation_tags]
+        operation_nodes_count = len(operation_nodes)
```
Sure.

potiuk replied:
Fixed in feeb75b855773880bcd9e0e921d2d30ed1bca290

potiuk commented on o2a/templates/prepare.tpl:
```diff
@@ -12,10 +12,16 @@
+{#
+ In BashOperator in Composer we can't read from $DAGS_FOLDER (~/dags) - permission denied.
+ However we can read from ~/data -> /home/airflow/gcs/data.
+ The easiest way to access it is using the $DAGS_FOLDER env variable.
```
Not sure. It's a left-over from the previous version. Removed.

potiuk replied:
Fixed in 0b1e2320ebc14f953732e8fa377b48b60d121461

potiuk commented:
Comments applied. If you are good with it, I rebase/squash and merge.
